src/isolation/filemap.rs

```diff
@@ -105,27 +123,111 @@ impl UhyveFileMap {
+	pub fn insert_fd_path(&mut self, fd: RawFd, guest_path: &str) {
+		// Don't insert standard streams (which "conflict" with Uhyve's).
+		if fd > 2 {
+			self.fdmap.insert(fd, guest_path.into());
+		} else {
+			warn!("Guest attempted to insert negative/standard stream {fd}, ignoring...")
+		}
+	}
+
+	/// Removes an fd from UhyveFileMap. This is only used by [crate::hypercall::close],
+	/// under the expectation that a new temporary file will be created if the guest
+	/// attempts to open a file of the same path again.
+	///
+	/// * `fd` - The file descriptor of the file being removed.
+	pub fn close_fd(&mut self, fd: RawFd) {
+		trace!("close_fd: {:#?}", &self.fdmap);
+		// The file descriptor in fdclosed is supposedly still alive.
+		// It is safe to assume that the host OS will not assign the
+		// same file descriptor to another opened file, until _after_
+		// the file has been closed.
+		if let Some(&fd) = self.unlinkedfd.get(&fd) {
+			self.unlinkedfd.remove(&fd);
+		} else {
+			self.remove_fd(fd);
+		}
+	}
+
+	/// Removes an fd from UhyveFileMap. This is only used by [crate::hypercall::close],
+	/// under the expectation that a new temporary file will be created if the guest
+	/// attempts to open a file of the same path again.
+	///
+	/// * `fd` - The file descriptor of the file being removed.
+	pub fn remove_fd(&mut self, fd: RawFd) {
+		trace!("remove_fd: {:#?}", &self.fdmap);
+		// Don't close Uhyve's standard streams.
+		if fd > 2 {
+			self.fdmap.remove_by_left(&fd);
+		} else {
+			warn!("Guest attempted to remove negative/standard stream {fd}, ignoring...")
+		}
+	}
+
+	/// Removes an entry (and its corresponding file descriptor) from UhyveFileMap.
+	///
 	/// Exclusively used by [crate::hypercall::unlink] for the event that
 	/// a file, which is mapped, is removed together with its corresponding
 	/// inode object. The intention is for Uhyve to create a new temporary
 	/// file, should the guest OS request to access the same guest path after
 	/// its corresponding host path has been unlinked. Otherwise, this would
 	/// prompt security mechanisms like Landlock to kill Uhyve.
 	///
-	/// * `guest_path` - The requested guest path.
-	pub fn remove_path(&mut self, guest_path: &str) {
-		// Returns None if the guest path is not mapped, i.e. a temporary file
-		// is unlinked or a parent directory has been mapped instead.
+	/// * `guest_path` - The path of the file being removed.
+	pub fn unlink_guest_path(&mut self, guest_path: &str) {
+		trace!("unlink_guest_path: {:#?}", &guest_path);
+		if let Some(fd) = self.fdmap.get_by_right(guest_path) {
+			self.unlinkedfd.insert(*fd);
+			self.fdmap.remove_by_right(guest_path);
+		}
 		self.files.remove(guest_path);
 	}
 }
 
+impl Drop for UhyveFileMap {
+	fn drop(&mut self) {
+		for fd in self.get_fds() {
+			// This creates a File instance from a non-closed fd.
+			// Rust "will then close" the file at the end of the unsafe block.
+			unsafe { std::fs::File::from_raw_fd(*fd) };
```

Comment:
In `fdmap`, I stored integers as some sort of form of micro-optimization, which is why I don't use `OwnedFd`. However, there's not much sense in creating an instance of `File`, which contains `OwnedFd` (therefore using `OwnedFd`'s `Drop` trait), when I could use `OwnedFd` directly.